panic when strarting kube-proxy if images is set to null in k0s config

### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Linux 5.15.63-0-virt #1-Alpine SMP Mon, 29 Aug 2022 07:17:28 +0000 x86_64 GNU/Linux
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.16.2
PRETTY_NAME="Alpine Linux v3.16"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://gitlab.alpinelinux.org/alpine/aports/-/issues"
```


### Version

v1.24.2+k0s.0-204-gf61c9b44c

### Sysinfo

<details><summary>`k0s sysinfo`</summary>

```text
Machine ID: "39c2127b5dbb78bde7dad9280f39509b81e8699e9f993bb98125abff2c24f9d5" (from machine) (pass)
Total memory: 983.9 MiB (warning: 1.0 GiB recommended)
Disk space available for /var/lib/k0s: 6.8 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 5.15.63-0-virt (pass)
  Max. file descriptors per process: current: 4096 / max: 4096 (warning: < 65536)
  Executable in path: modprobe: /sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 1 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (pass)
    cgroup controller "freezer": available (pass)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (pass)
  CONFIG_CGROUPS: Control Group support: no kernel config found (warning)
  CONFIG_NAMESPACES: Namespaces support: no kernel config found (warning)
  CONFIG_NET: Networking support: no kernel config found (warning)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: no kernel config found (warning)
  CONFIG_PROC_FS: /proc file system support: no kernel config found (warning)
```
</details>


### What happened?

When setting `images: null` in k0s config, k0s segfaults.

### Steps to reproduce

1. Put an invalid config into `/etc/k0s/k0s.yaml`
    ```yaml
    apiVersion: k0s.k0sproject.io/v1beta1
    kind: ClusterConfig
    spec:
      api:
        address: 10.70.123.36
        sans:
        - 10.70.123.36
        - 127.0.0.1
      images: null # <- this breaks it
    ```

2. Start the controller.

### Expected behavior

No segfault, clear error message. Probably even better: Identical behavior for `null` fields and absent fields.

### Actual behavior

When starting the controller, it chokes and segfaults when trying to start kube-proxy.

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x230bbc2]

goroutine 343 [running]:
github.com/k0sproject/k0s/pkg/component/controller.(*KubeProxy).getConfig(0x37737a0?, 0xc0008eab40)
        /go/src/github.com/k0sproject/k0s/pkg/component/controller/kubeproxy.go:109 +0xe2
github.com/k0sproject/k0s/pkg/component/controller.(*KubeProxy).Reconcile(0xc000d27500, {0x4?, 0x31192c7?}, 0xc0008eab40)
        /go/src/github.com/k0sproject/k0s/pkg/component/controller/kubeproxy.go:77 +0xc8
github.com/k0sproject/k0s/pkg/component.(*Manager).reconcileComponent(0xc0000e0680?, {0x379aa20, 0xc00109d680}, {0x37826a0?, 0xc000d27500}, 0x1?)
        /go/src/github.com/k0sproject/k0s/pkg/component/manager.go:166 +0x156
github.com/k0sproject/k0s/pkg/component.(*Manager).Reconcile(0xc0010c6a50, {0x379aa20, 0xc00109d680}, 0xc0008eab40)
        /go/src/github.com/k0sproject/k0s/pkg/component/manager.go:144 +0x254
github.com/k0sproject/k0s/pkg/component/controller.(*ClusterConfigReconciler).Start.func2()
        /go/src/github.com/k0sproject/k0s/pkg/component/controller/clusterConfig.go:153 +0x1cf
created by github.com/k0sproject/k0s/pkg/component/controller.(*ClusterConfigReconciler).Start
        /go/src/github.com/k0sproject/k0s/pkg/component/controller/clusterConfig.go:137 +0x165
```

### Screenshots and logs

_No response_

### Additional context

See #2133.

--- a/pkg/apis/k0s.k0sproject.io/v1beta1/clusterconfig_types.go
+++ b/pkg/apis/k0s.k0sproject.io/v1beta1/clusterconfig_types.go
@@ -215,7 +215,47 @@ func (c *ClusterConfig) UnmarshalJSON(data []byte) error {
 	decoder := json.NewDecoder(bytes.NewReader(data))
 	decoder.DisallowUnknownFields()
 
-	return decoder.Decode(jc)
+	err := decoder.Decode(jc)
+	if err != nil {
+		return err
+	}
+
+	if jc.Spec == nil {
+		jc.Spec = DefaultClusterSpec(storage)
+		return nil
+	}
+	if jc.Spec.Storage == nil {
+		jc.Spec.Storage = DefaultStorageSpec()
+	}
+	if jc.Spec.Extensions == nil {
+		jc.Spec.Extensions = DefaultExtensions()
+	}
+	if jc.Spec.Network == nil {
+		jc.Spec.Network = DefaultNetwork()
+	}
+	if jc.Spec.API == nil {
+		jc.Spec.API = DefaultAPISpec()
+	}
+	if jc.Spec.ControllerManager == nil {
+		jc.Spec.ControllerManager = DefaultControllerManagerSpec()
+	}
+	if jc.Spec.Scheduler == nil {
+		jc.Spec.Scheduler = DefaultSchedulerSpec()
+	}
+	if jc.Spec.Install == nil {
+		jc.Spec.Install = DefaultInstallSpec()
+	}
+	if jc.Spec.Images == nil {
+		jc.Spec.Images = DefaultClusterImages()
+	}
+	if jc.Spec.Telemetry == nil {
+		jc.Spec.Telemetry = DefaultClusterTelemetry()
+	}
+	if jc.Spec.Konnectivity == nil {
+		jc.Spec.Konnectivity = DefaultKonnectivitySpec()
+	}
+
+	return nil
 }
 
 // DefaultClusterSpec default settings

--- a/pkg/apis/k0s.k0sproject.io/v1beta1/clusterconfig_types_test.go
+++ b/pkg/apis/k0s.k0sproject.io/v1beta1/clusterconfig_types_test.go
@@ -167,6 +167,40 @@ spec:
 	assert.Equal(t, "https://1.2.3.4:9443", c.Spec.API.K0sControlPlaneAPIAddress())
 }
 
+func TestNullValues(t *testing.T) {
+	yamlData := `
+apiVersion: k0s.k0sproject.io/v1beta1
+kind: ClusterConfig
+metadata:
+  name: foobar
+spec:
+  images: null
+  storage: null
+  network: null
+  api: null
+  extensions: null
+  controllerManager: null
+  scheduler: null
+  installConfig: null
+  telemetry: null
+  konnectivity: null
+`
+
+	c, err := ConfigFromString(yamlData)
+	assert.NoError(t, err)
+	assert.Equal(t, DefaultClusterImages(), c.Spec.Images)
+	assert.Equal(t, DefaultStorageSpec(), c.Spec.Storage)
+	assert.Equal(t, DefaultNetwork(), c.Spec.Network)
+	assert.Equal(t, DefaultAPISpec(), c.Spec.API)
+	assert.Equal(t, DefaultExtensions(), c.Spec.Extensions)
+	assert.Equal(t, DefaultStorageSpec(), c.Spec.Storage)
+	assert.Equal(t, DefaultControllerManagerSpec(), c.Spec.ControllerManager)
+	assert.Equal(t, DefaultSchedulerSpec(), c.Spec.Scheduler)
+	assert.Equal(t, DefaultInstallSpec(), c.Spec.Install)
+	assert.Equal(t, DefaultClusterTelemetry(), c.Spec.Telemetry)
+	assert.Equal(t, DefaultKonnectivitySpec(), c.Spec.Konnectivity)
+}
+
 func TestWorkerProfileConfig(t *testing.T) {
 	yamlData := `
 apiVersion: k0s.k0sproject.io/v1beta1

